spark/src/main/scala/com/gu/tableversions/spark/VersionContext.scala

```diff
@@ -17,46 +17,32 @@ import org.apache.spark.sql.{Dataset, Row, SaveMode, SparkSession}
   * Code for writing Spark datasets to storage in a version-aware manner, taking in version information,
   * using the appropriate paths for storage, and committing version changes.
   */
-object VersionedDataset {
-
-  implicit class DatasetOps[T](val delegate: Dataset[T])(
-      implicit tableVersions: TableVersions[IO],
-      metastore: Metastore[IO],
-      generateVersion: IO[Version]) {
-
-    /**
-      * Insert the dataset into the given versioned table.
-      *
-      * This emulates the behaviour of Hive inserts in that it will overwrite any partitions present in the dataset,
-      * while leaving other partitions unchanged.
-      *
-      * @return a tuple containing the updated table version information, and a list of the changes that were applied
-      *         to the metastore.
-      */
-    def versionedInsertInto(table: TableDefinition, userId: UserId, message: String): (TableVersion, TableChanges) =
-      versionedInsertDatasetIntoTable(delegate, table, userId, message).unsafeRunSync()
-
-  }
+final case class VersionContext(
+    tableVersions: TableVersions[IO],
+    metastore: Metastore[IO],
+    generateVersion: IO[Version]
+) {
```

Comment:
This is a great change, just one thought: I wonder if it would make sense to move this `VersionContext` class into `core`? The `versionedInsertDatasetIntoTable` would then have to move out of it again, of course.

The reason I'm wondering about this is that this could make a good place to hang some functionality that interacts with both the `TableVersions` and the `Metastore`. For example the `TableLoader.checkout()` method: https://github.com/guardian/table-versions/blob/master/examples/src/main/scala/com/gu/tableversions/examples/TableLoader.scala#L47. This currently lives in example code, but it would be useful to have this somewhere in the real code. It doesn't belong in either of the two underlying services as it interacts with both. It also doesn't rely on Spark or any of the concrete implementations, so `core` would seem like a good place for it.

Likewise, the "commit and sync metastore" parts of the current `versionedInsertDatasetIntoTable` method could perhaps be pulled into common code int he same place.

What do you think?

Reply:
Yep, that sounds like a better place for it. Would you like to jump in and move things around? I think you'll have a better sense for exactly where to locate things.

It would be nice to preserve the ergonomics of enriching `Dataset`/`Dataframe` with a versioned insert method *without* having to pass the `VersionContext` around implicitly. I think I've got an idea brewing for this, but would need to play with it.

Reply:
OK, I can have a quick go and see what it looks like. (Looking forward to hearing about your cunning idea!)

Reply:
So I had a play but without much luck. One issue is that the `Metastore` trait is in the separate `metastore` module; thinking about it, I think this module should be removed and the code go in `core`, as these are implementation-neutral interfaces and types, just as the `TableVersions` trait and related code.

Also, the `VersionContext` you've created here is really a Spark version context, which makes a lot of sense, collecting the bits needed to write Spark code in an easy to use package. That's probably a separate concept from any lower level service that doesn't know about Spark, and is useful even if such a thing is added. So I think the changes here are fine as they are.